cache_test.go

```diff
@@ -798,3 +803,147 @@ func DisableTime(tb testing.TB) {
 		internal.SetTimeProvider(time.Now)
 	})
 }
+
+// The following tests flex various critical sections in the way glob collections are handled (along
+// with almost all the cache), to attempt to trigger a race condition. The tests must be run with
+// -race for this to have any use.
+func TestGlobRace(t *testing.T) {
+	prefix := globCollectionPrefix + "foo/"
+
+	// This tests many writers all competing for writes against overlapping entries.
+	t.Run("update", func(t *testing.T) {
+		c := newCache()
+
+		const (
+			entries = 100
+			writers = 100
+			count   = 100
+			readers = 100
+
+			doneVersion = "done"
+		)
+
+		entryNames := make([]string, entries)
+		for i := range entryNames {
+			name := prefix + strconv.Itoa(i)
+			entryNames[i] = name
+		}
+
+		var writesDone, readsDone sync.WaitGroup
+		writesDone.Add(writers)
+		readsDone.Add(writers * readers)
+
+		for range readers {
+			h := testutils.NewSubscriptionHandler(func(name string, r *ads.Resource[*Timestamp], _ ads.SubscriptionMetadata) {
+				if r.Version == doneVersion {
+					readsDone.Done()
+				}
+			})
+			c.Subscribe(ads.WildcardSubscription, h)
+		}
+
+		for range writers {
+			names := slices.Clone(entryNames)
+			shuffle := func() []string {
+				rand.Shuffle(len(names), func(i, j int) {
+					names[i], names[j] = names[j], names[i]
+				})
+				return names
+			}
+			go func() {
+				defer writesDone.Done()
+				for range count {
+					for _, name := range shuffle() {
+						c.Set(name, "1", new(Timestamp), time.Time{})
+					}
+				}
+			}()
+		}
+		go func() {
+			writesDone.Wait()
+			for _, name := range entryNames {
+				c.Set(name, doneVersion, new(Timestamp), time.Time{})
+			}
+		}()
+
+		readsDone.Wait()
+	})
+	// This tests subscribing to a collection that is currently being updated.
+	t.Run("subscribe", func(t *testing.T) {
```

Comment:
nit: "subscribe to collection being updated/cleared"

Reply:
Done